Source/Core/Common/MemArenaWin.cpp

```diff
@@ -1,71 +1,410 @@
+    // (it indicates a bug in the emulated memory mapping logic)
+    for (const auto& r : m_regions)
+    {
+      size_t mapped_region_count = 0;
+      if (r.m_is_mapped)
+        ++mapped_region_count;
+      if (mapped_region_count > 0)
+      {
+        PanicAlertFmt("Error while releasing fastmem region: {} regions are still mapped!",
+                      mapped_region_count);
+      }
+    }
+
+    // then free memory
+    VirtualFree(m_reserved_region, 0, MEM_RELEASE);
+    m_reserved_region = nullptr;
+    m_regions.clear();
+  }
+}
+
+WindowsMemoryRegion* MemArena::EnsureSplitRegionForMapping(void* start_address, size_t size)
+{
+  u8* const address = static_cast<u8*>(start_address);
+  auto& regions = m_regions;
+  if (regions.empty())
+  {
+    NOTICE_LOG_FMT(MEMMAP, "Tried to map a memory region without reserving a memory block first.");
+    return nullptr;
+  }
+
+  // find closest region that is <= the given address by using upper bound and decrementing
+  auto it = std::upper_bound(
+      regions.begin(), regions.end(), address,
+      [](u8* addr, const WindowsMemoryRegion& region) { return addr < region.m_start; });
+  if (it == regions.begin())
+  {
+    // this should never happen, implies that the given address is before the start of the
+    // reserved memory block
+    NOTICE_LOG_FMT(MEMMAP, "Invalid address {} given to map.", fmt::ptr(address));
+    return nullptr;
+  }
+  --it;
+
+  if (it->m_is_mapped)
+  {
+    NOTICE_LOG_FMT(MEMMAP,
+                   "Address to map {} with a size of 0x{:x} overlaps with existing mapping "
+                   "at {}.",
+                   fmt::ptr(address), size, fmt::ptr(it->m_start));
+    return nullptr;
+  }
+
+  const size_t mapping_index = it - regions.begin();
+  u8* const mapping_address = it->m_start;
+  const size_t mapping_size = it->m_size;
+  if (mapping_address == address)
+  {
+    // if this region is already split up correctly we don't have to do anything
+    if (mapping_size == size)
+      return &*it;
+
+    // if this region is smaller than the requested size we can't map
+    if (mapping_size < size)
+    {
+      NOTICE_LOG_FMT(MEMMAP,
+                     "Not enough free space at address {} to map 0x{:x} bytes (0x{:x} available).",
+                     fmt::ptr(mapping_address), size, mapping_size);
+      return nullptr;
+    }
+
+    // split region
+    if (!VirtualFree(address, size, MEM_RELEASE | MEM_PRESERVE_PLACEHOLDER))
+    {
+      NOTICE_LOG_FMT(MEMMAP, "Region splitting failed: {}", GetLastErrorString());
+      return nullptr;
+    }
+
+    // update tracked mappings and return the first of the two
+    it->m_size = size;
+    u8* const new_mapping_start = address + size;
+    const size_t new_mapping_size = mapping_size - size;
+    regions.insert(it + 1, WindowsMemoryRegion(new_mapping_start, new_mapping_size, false));
+    return &regions[mapping_index];
+  }
+
+  assert(mapping_address < address);
```

Comment:
This should use Dolphin's `ASSERT` macro in `Common/Assert.h` instead (and the include of `cassert` should be removed).